src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala

```diff
@@ -49,15 +49,67 @@ object JndiSchemaDAO {
   }
 }
 
-class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig)(implicit executionContext: ExecutionContext) extends JndiSupport with DirectorySubjectNameSupport {
+class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig, val schemaLockConfig: SchemaLockConfig)(implicit executionContext: ExecutionContext) extends JndiSupport with DirectorySubjectNameSupport with LazyLogging {
+
+  def initSchemaLock(): Future[Unit] = {
+    logger.info("Acquiring schema lock...")
+
+    for {
+      _ <- createSchemaChangeLockSchema()
+      _ <- createOrgUnit(schemaLockOu)
+      _ <- insertSchemaLock(schemaLockConfig.maxTimeToWait / schemaLockConfig.recheckTimeInterval)
+    } yield ()
+  }
 
   def init(): Future[Unit] = {
     for {
+      _ <- if(schemaLockConfig.lockSchemaOnBoot) initSchemaLock() else Future.successful(())
       _ <- destroySchema()
       _ <- createSchema()
+      _ <- if(schemaLockConfig.lockSchemaOnBoot) removeSchemaLock() else Future.successful(())
     } yield ()
   }
 
+
+  def insertSchemaLock(remainingAttempts: Int): Future[Unit] = withContext { ctx =>
+    if(remainingAttempts == 0) {
+      logger.error(s"Could not acquire schema lock after ${schemaLockConfig.maxTimeToWait} seconds...")
+      throw new Exception(s"Could not acquire schema lock after ${schemaLockConfig.maxTimeToWait} seconds...")
+    }
+    else {
+      val resourceContext = new BaseDirContext {
+        override def getAttributes(name: String): Attributes = {
+          val myAttrs = new BasicAttributes(true) // Case ignore
+
+          val oc = new BasicAttribute("objectclass")
+          Seq("top", "lockId").foreach(oc.add)
+          myAttrs.put(oc)
+          myAttrs.put("lockId", "0")
+          myAttrs.put("instanceId", schemaLockConfig.instanceId)
+
+          myAttrs
+        }
+      }
+
+      Try {
+        ctx.bind(schemaLockDn, resourceContext)
+      } match {
+        case Success(_) => logger.info("Schema lock acquired...")
+        case Failure(t: NameAlreadyBoundException) => {
+          logger.info(s"Schema is already locked. Waiting for schema lock...")
+          Thread.sleep(schemaLockConfig.recheckTimeInterval * 1000)
+          insertSchemaLock(remainingAttempts - 1)
+        }
+      }
+    }
+  }
+
+  def removeSchemaLock(): Future[Unit] = withContext { ctx =>
```

Comment:
We should have some documentation on how to manually remove the lock just in case something fails here. Maybe we should also retry this a few times to make it more unlikely that we'll have to manually intervene.

Reply:
Importantly, the doc needs to be easy to find by someone doing a release in case something goes wrong.